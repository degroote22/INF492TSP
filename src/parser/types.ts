interface IParsedBase {
  name: string;
  type: "TSP";
  comment: string;
  dimension: number;
}

export interface IParsedExplicit extends IParsedBase {
  edge_weight_type: "EXPLICIT";
  edge_weight_format: "LOWER_DIAG_ROW" | "UPPER_ROW";
  edge_weight_section: number[];
}

export interface IParsedEuc2D extends IParsedBase {
  edge_weight_type: "EUC_2D";
  node_coord_section: number[][];
}

export type IParsed = IParsedExplicit | IParsedEuc2D;
