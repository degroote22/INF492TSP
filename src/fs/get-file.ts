import * as fs from "fs";
export const getFile = (fileName: string) => {
  return fs.readFileSync(fileName, "utf-8");
};
