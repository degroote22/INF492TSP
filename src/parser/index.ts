const parseLines = (lines: RegExpMatchArray | null) => {
  if (!lines) {
    throw Error(`Arquivo com aparente defeito.`);
  }

  let capturingBlockEdgeWeightSection = false;
  let capturingBlockNodeCoordSection = false;
  const mapped = lines.reduce(
    (prev, curr) => {
      if (curr.replace(/ /g, "").length === 0 || curr.startsWith("EOF")) {
        return prev;
      }

      if (capturingBlockNodeCoordSection) {
        return {
          ...prev,
          ["NODE_COORD_SECTION".toLowerCase()]: [
            ...(prev["NODE_COORD_SECTION".toLowerCase()] || []),
            curr
              .split(" ")
              .filter(x => x.replace(/ /g, "").length !== 0)
              .map(x => Number(x))
          ]
        };
      }

      if (capturingBlockEdgeWeightSection) {
        return {
          ...prev,
          ["EDGE_WEIGHT_SECTION".toLowerCase()]: [
            ...(prev["EDGE_WEIGHT_SECTION".toLowerCase()] || []),
            ...curr
              .split(" ")
              .filter(x => x.length)
              .map(x => parseInt(x, 10))
          ]
        };
      }

      if (curr === "NODE_COORD_SECTION") {
        // parse block
        capturingBlockNodeCoordSection = true;
        return prev;
      } else if (curr === "EDGE_WEIGHT_SECTION") {
        // parse block
        capturingBlockEdgeWeightSection = true;
        return prev;
      } else {
        // parse line
        const parts = curr.replace(/ /g, "").split(":");
        const name = parts[0].toLowerCase();
        const value =
          name === "dimension"
            ? parseInt(parts[1], 10)
            : parts[1].split(" ")[0];
        return { ...prev, [name]: value };
      }
    },
    {} as { [key: string]: any }
  );

  return mapped as IParsed;
};

interface IParsedBase {
  name: string;
  type: "TSP";
  comment: string;
  dimension: number;
}

export interface IParsedExplicit extends IParsedBase {
  edge_weight_type: "EXPLICIT";
  edge_weight_format: "LOWER_DIAG_ROW" | "UPPER_ROW";
  edge_weight_section: number[];
}

export interface IParsedEuc2D extends IParsedBase {
  edge_weight_type: "EUC_2D";
  node_coord_section: number[][];
}

export type IParsed = IParsedExplicit | IParsedEuc2D;

export const parse = ({ file, type }: { file: string; type: "TSP" }) => {
  const arrayOfLines = file.match(/[^\r\n]+/g);

  const parsed = parseLines(arrayOfLines);
  if (!parsed.type.startsWith(type)) {
    throw Error("Tipo de arquivo n√£o suportado");
  }

  return parsed;
};
