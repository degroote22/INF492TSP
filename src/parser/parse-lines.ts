import { IParsed } from "./types";

function makeIntList(curr: string): number[] {
  return curr
    .split(" ")
    .filter(x => x.length)
    .map(x => parseInt(x, 10));
}

function makeNumberList(curr: string): number[] {
  return curr
    .split(" ")
    .filter(x => x.replace(/ /g, "").length !== 0)
    .map(x => Number(x));
}

function captureBlockLineEdgeWeightSection(
  prev: { [key: string]: any },
  curr: string
): { [key: string]: any } {
  return {
    ...prev,
    edge_weight_section: [
      ...(prev.edge_weight_section || []),
      ...makeIntList(curr)
    ]
  };
}

function captureBlockLineNodeCoordSection(
  prev: { [key: string]: any },
  curr: string
): { [key: string]: any } {
  return {
    ...prev,
    node_coord_section: [
      ...(prev.node_coord_section || []),
      makeNumberList(curr)
    ]
  };
}

function shouldIgnoreLine(curr: string): boolean {
  return curr.replace(/ /g, "").length === 0 || curr.startsWith("EOF");
}

function parseLine(curr: string, prev: { [key: string]: any }) {
  const parts = curr.replace(/ /g, "").split(":");
  const name = parts[0].toLowerCase();
  const value = name === "dimension" ? parseInt(parts[1], 10) : parts[1];
  return { ...prev, [name]: value };
}

export const parseLines = (lines: RegExpMatchArray | null) => {
  if (!lines) {
    throw Error(`Arquivo com aparente defeito.`);
  }

  let capturingBlockEdgeWeightSection = false;
  let capturingBlockNodeCoordSection = false;
  const mapped = lines.reduce(
    (prev, curr) => {
      if (shouldIgnoreLine(curr)) {
        return prev;
      }

      if (capturingBlockNodeCoordSection) {
        return captureBlockLineNodeCoordSection(prev, curr);
      }

      if (capturingBlockEdgeWeightSection) {
        return captureBlockLineEdgeWeightSection(prev, curr);
      }

      if (curr === "NODE_COORD_SECTION") {
        // parse block
        capturingBlockNodeCoordSection = true;
        return prev;
      } else if (curr === "EDGE_WEIGHT_SECTION") {
        // parse block
        capturingBlockEdgeWeightSection = true;
        return prev;
      } else {
        // parse line
        return parseLine(curr, prev);
      }
    },
    {} as { [key: string]: any }
  );

  return mapped as IParsed;
};
