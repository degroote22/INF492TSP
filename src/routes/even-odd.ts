import { range } from "../misc/utils";

export const evenOddRoute = (dimension: number) => {
  const arr = range(dimension).map(x => x + 1);
  let evens = [] as number[];

  let odds = [] as number[];

  arr.forEach(i => {
    if (i % 2 === 0) {
      evens.push(i);
    } else {
      odds.push(i);
    }
  });

  const sorted = [...odds, ...evens];

  return [...sorted, sorted[0]];
};
