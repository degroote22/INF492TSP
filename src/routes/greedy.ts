import { range, getDistance } from "../misc/utils";

const getNextCity = ({
  dimension,
  matrix,
  lastCity,
  route
}: {
  matrix: number[][];
  route: number[];
  lastCity: number;
  dimension: number;
}) => {
  const arr = range(dimension)
    .map((_, index) => index + 1)
    .filter(x => route.indexOf(x) === -1)
    .map(index => ({
      index,
      distance: getDistance({ c1: index, c2: lastCity, ldrMatrix: matrix })
    }))
    .sort((a, b) => a.distance - b.distance);
  return arr[0];
};

export const oneHeadRoute = ({
  matrix,
  dimension
}: {
  matrix: number[][];
  dimension: number;
}) => {
  let route = [1];

  for (let index = 0; index < dimension; index++) {
    const lastCity = route[route.length - 1];

    const nextCity = getNextCity({ matrix, route, dimension, lastCity });
    if (nextCity) {
      route.push(nextCity.index);
    } else {
      route.push(1);
    }
  }

  return route;
};

export const twoHeadRoute = ({
  matrix,
  dimension
}: {
  matrix: number[][];
  dimension: number;
}) => {
  let route = [1];

  for (let index = 0; index < dimension; index++) {
    const lastCityHead1 = route[0];
    const lastCityHead2 = route[route.length - 1];

    const nextCityHead1 = getNextCity({
      matrix,
      route,
      dimension,
      lastCity: lastCityHead1
    });
    const nextCityHead2 = getNextCity({
      matrix,
      route,
      dimension,
      lastCity: lastCityHead2
    });

    if (!nextCityHead1 && !nextCityHead2) {
      // último item da iteração
      // nunca há o caso de ter um e não ter o outro
      // porque os dois retornarão vazios só quando o filter
      // do getNextCity (que ignora a rota atual) chegar no final

      // lastCityHead1 é o primeiro do array
      // isso acontece na iteração final
      route.push(lastCityHead1);
    } else if (nextCityHead1.distance < nextCityHead2.distance) {
      // adiciona na frente do array
      route = [nextCityHead1.index, ...route];
    } else {
      // adiciona no fim do array
      route.push(nextCityHead2.index);
    }
  }

  return route;
};
