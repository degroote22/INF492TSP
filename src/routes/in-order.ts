import { range } from "../misc/utils";

export const inOrderRoute = (dimension: number) => {
  const arr = range(dimension).map(x => x + 1);

  return [...arr, arr[0]];
};
