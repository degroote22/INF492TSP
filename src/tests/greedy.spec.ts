import { oneHeadRoute, twoHeadRoute } from "../routes/greedy";
import { getFile } from "../fs/get-file";
import { parse } from "../parser";

describe.only("greety solver", () => {
  const cases = [
    { n: "data/gr17.tsp" },
    { n: "data/gr21.tsp" },
    { n: "data/gr48.tsp" },
    { n: "data/gr120.tsp" },
    { n: "data/brazil58.tsp" },
    { n: "data/berlin52.tsp" },
    { n: "data/fri26.tsp" },
    { n: "data/eil51.tsp" },
    { n: "data/eil76.tsp" },
    { n: "data/eil101.tsp" },
    { n: "data/a280.tsp" }
  ];

  cases.forEach(({ n }) => {
    test(n, () => {
      const file = getFile(n);
      const parsed = parse({ file, type: "TSP" });
      const orderOneHead = oneHeadRoute(parsed);
      const orderTwoHead = twoHeadRoute(parsed);
      expect(orderOneHead).toMatchSnapshot();
      expect(orderTwoHead).toMatchSnapshot();
    });
  });
});
