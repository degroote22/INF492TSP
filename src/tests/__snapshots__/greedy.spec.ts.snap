// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`greety solver data/a280.tsp 1`] = `
Array [
  1,
  280,
  2,
  3,
  279,
  278,
  4,
  277,
  276,
  275,
  274,
  273,
  272,
  271,
  16,
  17,
  18,
  133,
  132,
  19,
  20,
  131,
  130,
  129,
  128,
  127,
  126,
  125,
  30,
  31,
  32,
  29,
  28,
  27,
  26,
  22,
  25,
  23,
  24,
  14,
  13,
  12,
  11,
  10,
  8,
  9,
  7,
  6,
  5,
  260,
  261,
  262,
  263,
  264,
  265,
  266,
  140,
  141,
  148,
  139,
  138,
  137,
  267,
  268,
  136,
  135,
  269,
  270,
  134,
  21,
  154,
  155,
  153,
  156,
  152,
  151,
  177,
  176,
  181,
  182,
  183,
  184,
  185,
  187,
  186,
  189,
  188,
  190,
  191,
  192,
  194,
  195,
  196,
  197,
  198,
  201,
  200,
  144,
  143,
  146,
  147,
  142,
  145,
  199,
  202,
  203,
  204,
  205,
  206,
  208,
  253,
  254,
  255,
  256,
  257,
  258,
  259,
  249,
  250,
  251,
  230,
  229,
  228,
  227,
  226,
  225,
  224,
  223,
  222,
  221,
  220,
  219,
  218,
  217,
  216,
  213,
  212,
  207,
  210,
  211,
  214,
  215,
  209,
  252,
  248,
  247,
  244,
  241,
  240,
  239,
  246,
  245,
  231,
  238,
  237,
  236,
  235,
  234,
  233,
  232,
  243,
  242,
  15,
  149,
  150,
  178,
  179,
  180,
  175,
  160,
  159,
  158,
  157,
  119,
  120,
  121,
  122,
  123,
  124,
  34,
  33,
  36,
  35,
  38,
  37,
  39,
  40,
  41,
  42,
  43,
  60,
  61,
  118,
  117,
  115,
  114,
  113,
  87,
  84,
  85,
  65,
  66,
  64,
  63,
  62,
  116,
  86,
  83,
  82,
  81,
  80,
  79,
  76,
  75,
  77,
  78,
  95,
  96,
  97,
  94,
  93,
  98,
  99,
  100,
  101,
  102,
  103,
  104,
  108,
  110,
  111,
  112,
  88,
  89,
  109,
  90,
  91,
  92,
  169,
  170,
  172,
  171,
  167,
  168,
  166,
  165,
  164,
  163,
  162,
  161,
  174,
  173,
  106,
  105,
  107,
  74,
  73,
  72,
  71,
  70,
  67,
  68,
  69,
  58,
  57,
  44,
  45,
  46,
  47,
  48,
  49,
  50,
  51,
  52,
  53,
  54,
  55,
  56,
  59,
  193,
  1,
]
`;

exports[`greety solver data/a280.tsp 2`] = `
Array [
  193,
  192,
  191,
  188,
  189,
  190,
  186,
  187,
  185,
  184,
  183,
  182,
  181,
  180,
  179,
  146,
  143,
  201,
  196,
  195,
  194,
  197,
  198,
  199,
  145,
  144,
  200,
  202,
  203,
  204,
  205,
  206,
  226,
  225,
  224,
  223,
  221,
  222,
  219,
  220,
  217,
  218,
  215,
  216,
  213,
  214,
  210,
  211,
  212,
  207,
  208,
  209,
  252,
  253,
  254,
  257,
  256,
  249,
  248,
  247,
  250,
  251,
  230,
  229,
  228,
  227,
  235,
  234,
  232,
  233,
  236,
  237,
  238,
  231,
  246,
  239,
  241,
  240,
  245,
  244,
  243,
  242,
  1,
  280,
  2,
  3,
  279,
  278,
  4,
  277,
  276,
  275,
  274,
  273,
  272,
  271,
  16,
  17,
  18,
  133,
  132,
  19,
  20,
  131,
  130,
  129,
  128,
  127,
  126,
  125,
  30,
  31,
  32,
  29,
  28,
  27,
  26,
  22,
  25,
  23,
  24,
  14,
  13,
  12,
  11,
  10,
  8,
  9,
  7,
  6,
  5,
  260,
  261,
  262,
  263,
  264,
  265,
  266,
  140,
  141,
  142,
  147,
  148,
  139,
  138,
  137,
  136,
  268,
  267,
  269,
  270,
  134,
  135,
  149,
  150,
  178,
  151,
  152,
  156,
  153,
  155,
  154,
  122,
  123,
  124,
  121,
  120,
  119,
  157,
  158,
  159,
  160,
  175,
  161,
  162,
  163,
  164,
  165,
  166,
  167,
  168,
  169,
  170,
  171,
  172,
  173,
  106,
  105,
  104,
  108,
  110,
  111,
  114,
  115,
  117,
  118,
  61,
  60,
  43,
  42,
  41,
  40,
  39,
  38,
  37,
  36,
  35,
  34,
  33,
  49,
  52,
  51,
  50,
  48,
  53,
  54,
  55,
  56,
  45,
  44,
  59,
  58,
  57,
  46,
  47,
  68,
  69,
  70,
  67,
  66,
  65,
  85,
  84,
  87,
  113,
  116,
  86,
  63,
  62,
  64,
  71,
  72,
  73,
  74,
  75,
  76,
  77,
  78,
  79,
  81,
  82,
  83,
  88,
  112,
  109,
  89,
  90,
  91,
  92,
  102,
  103,
  101,
  100,
  99,
  98,
  97,
  96,
  95,
  94,
  93,
  80,
  107,
  174,
  176,
  177,
  21,
  15,
  259,
  258,
  255,
  193,
]
`;

exports[`greety solver data/berlin52.tsp 1`] = `
Array [
  1,
  22,
  49,
  32,
  36,
  35,
  34,
  39,
  40,
  38,
  37,
  48,
  24,
  5,
  15,
  6,
  4,
  25,
  46,
  44,
  16,
  50,
  20,
  23,
  31,
  18,
  3,
  19,
  45,
  41,
  8,
  10,
  9,
  43,
  33,
  51,
  12,
  28,
  27,
  26,
  47,
  13,
  14,
  52,
  11,
  29,
  30,
  21,
  17,
  42,
  7,
  2,
  1,
]
`;

exports[`greety solver data/berlin52.tsp 2`] = `
Array [
  29,
  2,
  7,
  42,
  17,
  21,
  30,
  23,
  20,
  50,
  16,
  44,
  46,
  25,
  4,
  6,
  15,
  5,
  24,
  48,
  37,
  38,
  40,
  39,
  34,
  35,
  36,
  32,
  49,
  1,
  22,
  31,
  18,
  3,
  19,
  45,
  41,
  8,
  10,
  9,
  43,
  33,
  51,
  12,
  28,
  27,
  26,
  47,
  13,
  14,
  52,
  11,
  29,
]
`;

exports[`greety solver data/brazil58.tsp 1`] = `
Array [
  1,
  35,
  15,
  41,
  58,
  29,
  46,
  19,
  34,
  50,
  27,
  39,
  6,
  9,
  55,
  37,
  51,
  4,
  5,
  57,
  18,
  24,
  49,
  7,
  8,
  12,
  10,
  56,
  11,
  20,
  30,
  23,
  42,
  33,
  53,
  40,
  31,
  45,
  22,
  25,
  13,
  3,
  14,
  26,
  44,
  38,
  48,
  43,
  21,
  17,
  28,
  52,
  54,
  32,
  16,
  47,
  2,
  36,
  1,
]
`;

exports[`greety solver data/brazil58.tsp 2`] = `
Array [
  52,
  28,
  17,
  21,
  51,
  37,
  43,
  48,
  38,
  44,
  26,
  14,
  3,
  13,
  25,
  22,
  45,
  31,
  40,
  53,
  33,
  42,
  23,
  30,
  20,
  11,
  56,
  10,
  12,
  8,
  7,
  49,
  24,
  18,
  57,
  4,
  5,
  36,
  2,
  47,
  16,
  32,
  54,
  1,
  35,
  15,
  41,
  58,
  29,
  46,
  19,
  34,
  50,
  27,
  39,
  6,
  9,
  55,
  52,
]
`;

exports[`greety solver data/eil51.tsp 1`] = `
Array [
  1,
  32,
  11,
  38,
  5,
  49,
  9,
  50,
  16,
  2,
  29,
  21,
  34,
  30,
  10,
  39,
  33,
  45,
  15,
  44,
  37,
  17,
  4,
  18,
  47,
  12,
  46,
  51,
  27,
  48,
  6,
  14,
  25,
  13,
  41,
  19,
  42,
  40,
  24,
  23,
  7,
  26,
  8,
  31,
  28,
  3,
  20,
  35,
  36,
  22,
  43,
  1,
]
`;

exports[`greety solver data/eil51.tsp 2`] = `
Array [
  40,
  43,
  24,
  6,
  17,
  37,
  33,
  45,
  15,
  44,
  42,
  19,
  41,
  13,
  25,
  14,
  18,
  4,
  47,
  12,
  46,
  51,
  27,
  48,
  23,
  7,
  26,
  8,
  31,
  28,
  22,
  1,
  32,
  11,
  38,
  5,
  49,
  9,
  50,
  16,
  2,
  29,
  21,
  34,
  30,
  10,
  39,
  20,
  35,
  36,
  3,
  40,
]
`;

exports[`greety solver data/eil76.tsp 1`] = `
Array [
  1,
  73,
  62,
  28,
  74,
  30,
  2,
  68,
  75,
  76,
  67,
  34,
  46,
  8,
  35,
  7,
  53,
  14,
  19,
  54,
  13,
  27,
  52,
  45,
  29,
  48,
  47,
  21,
  36,
  71,
  60,
  70,
  20,
  37,
  5,
  15,
  57,
  4,
  26,
  12,
  40,
  17,
  51,
  6,
  33,
  63,
  16,
  3,
  44,
  32,
  9,
  39,
  72,
  58,
  10,
  38,
  65,
  11,
  66,
  59,
  31,
  25,
  50,
  18,
  24,
  49,
  23,
  56,
  41,
  42,
  43,
  22,
  61,
  69,
  64,
  55,
  1,
]
`;

exports[`greety solver data/eil76.tsp 2`] = `
Array [
  69,
  61,
  22,
  64,
  42,
  41,
  43,
  1,
  73,
  62,
  28,
  74,
  30,
  2,
  68,
  75,
  76,
  67,
  34,
  46,
  8,
  35,
  7,
  53,
  14,
  19,
  54,
  13,
  27,
  52,
  45,
  29,
  48,
  47,
  21,
  36,
  71,
  60,
  70,
  20,
  37,
  5,
  15,
  57,
  4,
  26,
  12,
  40,
  17,
  51,
  6,
  33,
  63,
  16,
  3,
  44,
  32,
  9,
  39,
  72,
  58,
  10,
  38,
  65,
  11,
  66,
  59,
  31,
  25,
  50,
  18,
  24,
  49,
  23,
  56,
  55,
  69,
]
`;

exports[`greety solver data/eil101.tsp 1`] = `
Array [
  1,
  69,
  27,
  101,
  53,
  58,
  40,
  21,
  73,
  72,
  74,
  22,
  75,
  56,
  39,
  23,
  67,
  25,
  55,
  54,
  80,
  68,
  77,
  3,
  79,
  78,
  34,
  81,
  33,
  51,
  9,
  71,
  35,
  65,
  66,
  20,
  30,
  70,
  31,
  88,
  7,
  82,
  48,
  47,
  36,
  49,
  19,
  11,
  62,
  10,
  90,
  63,
  32,
  64,
  46,
  8,
  45,
  83,
  60,
  5,
  84,
  17,
  61,
  16,
  91,
  100,
  37,
  98,
  93,
  99,
  96,
  59,
  92,
  97,
  95,
  94,
  6,
  89,
  18,
  52,
  28,
  26,
  12,
  76,
  50,
  29,
  24,
  4,
  41,
  57,
  2,
  87,
  42,
  43,
  15,
  14,
  44,
  86,
  85,
  13,
  38,
  1,
]
`;

exports[`greety solver data/eil101.tsp 2`] = `
Array [
  64,
  49,
  36,
  46,
  8,
  45,
  17,
  84,
  5,
  60,
  83,
  18,
  52,
  7,
  82,
  48,
  47,
  19,
  11,
  63,
  32,
  90,
  10,
  62,
  88,
  31,
  70,
  30,
  20,
  66,
  65,
  35,
  71,
  9,
  51,
  33,
  81,
  34,
  78,
  79,
  3,
  77,
  76,
  50,
  1,
  69,
  27,
  101,
  53,
  58,
  40,
  21,
  73,
  72,
  74,
  22,
  75,
  56,
  39,
  23,
  67,
  25,
  55,
  54,
  80,
  68,
  12,
  26,
  28,
  89,
  6,
  94,
  95,
  97,
  92,
  59,
  99,
  96,
  93,
  85,
  98,
  37,
  100,
  91,
  16,
  61,
  86,
  44,
  14,
  42,
  87,
  2,
  57,
  15,
  43,
  38,
  13,
  41,
  4,
  24,
  29,
  64,
]
`;

exports[`greety solver data/fri26.tsp 1`] = `
Array [
  1,
  15,
  14,
  10,
  11,
  13,
  12,
  9,
  7,
  8,
  16,
  19,
  20,
  18,
  17,
  21,
  22,
  26,
  23,
  24,
  25,
  3,
  2,
  4,
  5,
  6,
  1,
]
`;

exports[`greety solver data/fri26.tsp 2`] = `
Array [
  6,
  5,
  4,
  3,
  2,
  1,
  15,
  14,
  10,
  11,
  13,
  12,
  9,
  7,
  8,
  16,
  19,
  20,
  18,
  17,
  21,
  22,
  26,
  23,
  24,
  25,
  6,
]
`;

exports[`greety solver data/gr17.tsp 1`] = `
Array [
  1,
  13,
  4,
  7,
  17,
  6,
  8,
  14,
  15,
  3,
  11,
  5,
  10,
  2,
  9,
  12,
  16,
  1,
]
`;

exports[`greety solver data/gr17.tsp 2`] = `
Array [
  9,
  12,
  16,
  1,
  13,
  4,
  7,
  17,
  6,
  8,
  14,
  15,
  3,
  11,
  5,
  10,
  2,
  9,
]
`;

exports[`greety solver data/gr21.tsp 1`] = `
Array [
  1,
  12,
  4,
  11,
  20,
  21,
  15,
  2,
  14,
  13,
  18,
  10,
  17,
  19,
  7,
  8,
  6,
  16,
  5,
  9,
  3,
  1,
]
`;

exports[`greety solver data/gr21.tsp 2`] = `
Array [
  3,
  9,
  5,
  16,
  6,
  8,
  7,
  1,
  12,
  4,
  11,
  20,
  10,
  18,
  21,
  15,
  2,
  14,
  13,
  17,
  19,
  3,
]
`;

exports[`greety solver data/gr48.tsp 1`] = `
Array [
  1,
  29,
  7,
  28,
  46,
  18,
  34,
  23,
  25,
  3,
  19,
  4,
  30,
  45,
  2,
  35,
  20,
  38,
  43,
  47,
  37,
  40,
  39,
  42,
  24,
  10,
  12,
  31,
  33,
  8,
  22,
  6,
  36,
  11,
  16,
  13,
  48,
  26,
  15,
  5,
  21,
  32,
  27,
  17,
  9,
  14,
  44,
  41,
  1,
]
`;

exports[`greety solver data/gr48.tsp 2`] = `
Array [
  44,
  41,
  24,
  15,
  26,
  14,
  9,
  17,
  27,
  32,
  21,
  5,
  12,
  10,
  31,
  33,
  8,
  22,
  6,
  36,
  16,
  11,
  48,
  13,
  1,
  29,
  7,
  28,
  46,
  18,
  34,
  23,
  25,
  3,
  19,
  4,
  30,
  45,
  2,
  35,
  20,
  38,
  43,
  47,
  37,
  40,
  39,
  42,
  44,
]
`;

exports[`greety solver data/gr120.tsp 1`] = `
Array [
  1,
  76,
  29,
  30,
  32,
  92,
  28,
  120,
  61,
  16,
  60,
  24,
  111,
  96,
  90,
  54,
  8,
  70,
  116,
  34,
  26,
  4,
  119,
  103,
  9,
  23,
  11,
  51,
  115,
  2,
  82,
  3,
  80,
  27,
  5,
  63,
  77,
  64,
  109,
  93,
  21,
  88,
  97,
  12,
  95,
  39,
  83,
  57,
  73,
  114,
  101,
  102,
  48,
  110,
  112,
  36,
  84,
  6,
  55,
  89,
  47,
  71,
  40,
  72,
  105,
  74,
  87,
  14,
  75,
  44,
  46,
  50,
  98,
  42,
  17,
  118,
  49,
  20,
  107,
  69,
  65,
  43,
  108,
  25,
  19,
  18,
  85,
  22,
  66,
  31,
  117,
  86,
  94,
  78,
  45,
  81,
  15,
  59,
  56,
  41,
  7,
  38,
  53,
  67,
  37,
  62,
  99,
  104,
  35,
  10,
  106,
  13,
  113,
  68,
  79,
  58,
  100,
  33,
  52,
  91,
  1,
]
`;

exports[`greety solver data/gr120.tsp 2`] = `
Array [
  13,
  91,
  52,
  33,
  100,
  58,
  79,
  68,
  113,
  1,
  76,
  29,
  30,
  32,
  92,
  28,
  120,
  61,
  16,
  60,
  24,
  111,
  96,
  90,
  54,
  8,
  70,
  116,
  34,
  26,
  4,
  119,
  103,
  9,
  23,
  11,
  51,
  115,
  2,
  82,
  3,
  80,
  27,
  5,
  63,
  77,
  64,
  109,
  93,
  21,
  88,
  97,
  12,
  95,
  39,
  83,
  57,
  73,
  114,
  101,
  102,
  48,
  110,
  112,
  36,
  84,
  6,
  55,
  89,
  47,
  71,
  40,
  72,
  105,
  74,
  87,
  14,
  75,
  44,
  46,
  50,
  98,
  42,
  17,
  118,
  49,
  20,
  107,
  69,
  65,
  43,
  108,
  25,
  19,
  18,
  85,
  22,
  66,
  31,
  117,
  86,
  94,
  78,
  45,
  81,
  15,
  59,
  56,
  41,
  7,
  38,
  53,
  67,
  37,
  62,
  99,
  104,
  35,
  10,
  106,
  13,
]
`;
